apps/settings/js/modules/base/module.js

```diff
@@ -0,0 +1,150 @@
+/**
+ * Module provides methods to create a new module. Any module created here has
+ * the ability of being extended by other modules. Existing properties are
+ * overridden.
+ *
+ * Creating a module
+ * @example
+ *   var NewModule = Module.create(function() {
+ *     this.value = 100;
+ *   });
+ *   NewModule.prototype.print = function() { console.log(this.value); };
+ *
+ *   var instance = NewModule();
+ *   instance.print(); // 100
+ *
+ * Extending a module
+ * @example
+ *   var AnotherModule = Module.create();
+ *   AnotherModule.prototype.inc = function() { this.value++; };
+ *   var ExtendingModule = NewModule.extend(AnotherModule);
+ *
+ *   var instance = ExtendingModule();
+ *   instance.inc();
+ *   instance.print(); // 101
+ *
+ * @module modules/base/module
+ */
+define(function() {
+  'use strict';
+
+  var LOG_LEVEL = {
+    NONE: 0,
+    TRACE: 1,
+    DEBUG: 2,
+    INFO: 3,
+    WARN: 4,
+    ERROR: 5,
+    ALL: 6
+  };
+
+  var _constructorMap = (function() {
+    var _ctorMap = new Map();
+    var _getConstructor = function(Module) {
+      return _ctorMap.get(Module) || function() {};
+    };
+    var _registerConstructor = function(Module, constructor) {
+      _ctorMap.set(Module, constructor || function() {});
+    };
+    return {
+      getConstructor: _getConstructor,
+      registerConstructor: _registerConstructor
+    };
+  })();
+
+  var _emptyFunction = function() {};
+  var _createLogger = function(name) {
+    switch (name) {
+      case 'TRACE':
+      case 'DEBUG':
+      case 'INFO':
+        return function(msg) {
+          console.log(name + ' ' + this._msgPrefix + msg);
+        };
+      case 'WARN':
+        return function(msg) {
+          console.warn(name + ' ' + this._msgPrefix + msg);
+        };
+      case 'ERROR':
+        return function(msg) {
+          console.error(name + ' ' + this._msgPrefix + msg);
+        };
+    }
+  };
+
+  var ModulePrototype = {
+    get _msgPrefix() {
+      return '[' + this.$name + ']: ';
+    },
+    trace: _emptyFunction,
+    debug: _emptyFunction,
+    info: _emptyFunction,
+    warn: _emptyFunction,
+    error: _emptyFunction,
+    throw: function(msg) {
+      throw new Error(this._msgPrefix + msg);
+    },
+    set _logLevel(value) {
+      Object.keys(LOG_LEVEL).forEach((name) => {
+        var level = LOG_LEVEL[name];
+        if (value >= level && value > 0) {
+          this[name.toLowerCase()] = _createLogger(name);
+        } else {
+          this[name.toLowerCase()] = _emptyFunction;
```

Comment:
Because we already set all these helper function to "_emptyFunction" at first, I don't think it is necessary to iterate all of them and set the others to "_emptyFunctions" again. Because normally, this is debug used code, we won't dynamically change the log level at runtime.